src/timely-util/src/builder_async.rs

```diff
@@ -117,53 +105,15 @@ impl<'handle, T: Timestamp, D: Container, P: Pull<BundleCore<T, D>>> Future
-                // to call `AsyncInputHandle::next_sync`, with no auto-deref happening, whose
-                // function signature is:
-                // for<'a> fn(&'a mut AsyncInputHandle) -> Option<(CapabilityRef<'a, T>, RefOrMut<'a, D>)>
-                //
-                // Therefore, by calling it with our `'handle` lifetimed handle we get back
-                // `'handle` lifetimed capability and data.
-                //
-                // An additional argument for safety is that the current borrow checker is happy
-                // accepting the body of this function without transmutes when either the Some or
-                // None branch is replaced by a panic!(). Therefore each individual path is safe on
-                // its own and in the full version we only ever call one of them, therefore the
-                // composition is also safe.
-                //
-                // ## Safety of ManuallyDrop
-                // `next_sync` wraps the returned `CapabilityRef` into a `ManuallyDrop` in order to
-                // not interfere with the borrow checker that will otherwise infer that `cap` may
-                // live until the end of this function and doesn't compile even when transmuting.
-                //
-                // ## Safety of transmute_copy
-                //
-                // We are forced to use `transmute_copy` and not a simple transmute because the
-                // compiler cannot understand that the two types have the same size, even though
-                // they clearly do, since they only differ in their lifetime. `transmute_copy` will
-                // create a bit copy of the capability and at the same time remove the wrapping
-                // `ManuallyDrop`. This is safe because `ManuallyDrop` is marked as
-                // `#[repr(transparent)]`, therefore it has the same layout as the T it wraps.
-                // Finally we don't need to care about running mem::forget() on the original cap
-                // that we created a copy out because it is wrapped in `ManuallyDrop` so no
-                // destructor will run.
-                //
-                let cap = unsafe {
-                    transmute_copy::<ManuallyDrop<CapabilityRef<'_, T>>, CapabilityRef<'handle, T>>(
-                        &cap,
-                    )
-                };
-                let data = unsafe { transmute::<RefOrMut<'_, D>, RefOrMut<'handle, D>>(data) };
 
-                Poll::Ready(Some(Event::Data(cap, data)))
-            }
-            None => {
+        type NextHTB<T, D> =
+            dyn for<'lt> WithLifetime<'lt, T = (CapabilityRef<'lt, T>, RefOrMut<'lt, D>)>;
+
+        match polonius::<NextHTB<T, D>, _, _, _>(handle, |handle| {
+            handle.sync_handle.next().ok_or(())
+        }) {
+            Ok((cap, data)) => Poll::Ready(Some(Event::Data(cap, data))),
+            Err((handle, _)) => {
```

Comment:
I would link to the docs, or at least give a lil explanation here

Reply:
Added!